appservice/src/deploy/localGitDeploy.ts

```diff
@@ -46,16 +48,40 @@ export async function localGitDeploy(client: SiteClient, fsPath: string): Promis
             return undefined;
             // tslint:disable-next-line:no-unsafe-any
         } else if (err.message.indexOf('error: failed to push') >= 0) {
-            const forcePush: vscode.MessageItem = { title: localize('forcePush', 'Force Push') };
+            const forcePushMessage: vscode.MessageItem = { title: localize('forcePush', 'Force Push') };
             const pushReject: string = localize('localGitPush', 'Push rejected due to Git history diverging.');
-            await ext.ui.showWarningMessage(pushReject, forcePush, DialogResponses.cancel);
-            // tslint:disable-next-line:no-floating-promises
-            localGit.push(remote, 'HEAD:master', { '-f': true });
+
+            if (await ext.ui.showWarningMessage(pushReject, forcePushMessage, DialogResponses.cancel) === forcePushMessage) {
+                await pushAndWaitForDeploymentToComplete(true);
+            } else {
+                throw new UserCancelledError();
+            }
         } else {
             throw err;
         }
     }
 
-    ext.outputChannel.appendLine(formatDeployLog(client, (localize('localGitDeploy', `Deploying Local Git repository to "${client.fullName}"...`))));
-    await waitForDeploymentToComplete(client, kuduClient, commitId);
+    async function pushAndWaitForDeploymentToComplete(forcePush: boolean = false): Promise<void> {
+        const tokenSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+        const token: vscode.CancellationToken = tokenSource.token;
+
+        await new Promise((resolve: () => void, reject: (error: Error) => void): void => {
+            // for whatever reason, is '-f' exists, true or false, it still force pushes
+            const pushOptions: git.Options = forcePush ? {'-f': true} : {};
+
+            localGit.push(remote, 'HEAD:master', pushOptions).catch(async (error) => {
+                tokenSource.cancel();
+                // tslint:disable-next-line:no-unsafe-any
+                reject(error);
+            });
+
+            waitForDeploymentToComplete(client, kuduClient, commitId, token).then(() => {
+                if (!token.isCancellationRequested) {
+                    resolve();
+                }
+            }).catch(reject);
+        });
+
+        tokenSource.dispose();
```

Comment:
Almost any time you do `dispose` it should be in a try/finally block so that it happens even on error